### PR TITLE
WIP API: Support for explicit Event timestamps and bulk Events

### LINKED ISSUE #76: API: Convert Event interface to concrete type
see this [comment](https://github.com/open-telemetry/opentelemetry-go/pull/65/files#r310237999)

Event is currently an interface. It is desirable that this be converted to a concrete Event type with support for lazy loading.

### PR DESCRIPTION
**Pull request by agneum**
Resolve #76

### CODE REVIEW

**Comment by agneum** [comment]
Is it perfect or should I add anything else?

File: api/event/event.go [hunk 1]
```diff
@@ -18,12 +18,18 @@ import (
...
+// Message retrieves message string of the Event.
+func (e *Event) Message() string {
+	return e.message
+}
 
-	// Attributes interface returns a copy of attributes associated with the Event.
-	Attributes() []core.KeyValue
+// Attributes returns a copy of attributes associated with the Event.
+func (e *Event) Attributes() []core.KeyValue {
+	return e.attr
```
Comment (rghetia) [inline api/event/event.go]: This should be a copy of attrs so the Event is immutable.

Comment (agneum) [reply api/event/event.go]: Done

**Comment by agneum** [comment]
> I don't feel we need an Event interface or a struct unless it's to support the unusual cases (explicit timestamp and/or bulk events, which OpenTracing did support). I'm requesting to update this PR with the API in PR99.

Should I wait until PR99 is merged into the master branch?

**Comment by agneum** [comment]
> I don't feel we need an Event interface or a struct unless it's to support the unusual cases (explicit timestamp and/or bulk events, which OpenTracing did support). I'm requesting to update this PR with the API in PR99.

I've added two methods to the Span interface: AddEventWithTimestamp and AddBulkEvent. How do you think, should I improve or change something?

**Comment by jmacd** [comment]
@agneum I updated the PR description, since it's changed a bit. I do not object to bulk events, but this is not something we have in the specification. Would you consider: (1) justifying your use-case, (2) separating the timestamped event stuff? 

(@krnowak has a similar change mixed up in another PR, we'd definitely fast-track the timestamped event support)

**Comment by agneum** [comment]
@jmacd I assumed that (1) it might need to generate some events, and then add them to the span at once. I supposed (2) it would be good if the timestamp contains the time the event was created, but not the time when it was added to the span. I checked the specification(https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-tracing.md) and did not see the requirements for the bulk operation. 
Anyway, if this feature is not very important, I can close the PR and help with more useful issues.

**Comment by jmacd** [comment]
Thank you @agneum. We agree to skip bulk event support for now, and I believe an event-with-timestamp change has already merged. We'll close this since you agreed.

**Comment by jmacd** [comment]
FYI the with-timestamp option arrives in this PR, which is approved but not merged:
https://github.com/open-telemetry/opentelemetry-go/pull/98

**Comment by agneum** [comment]
Thank you. I'll check this PR.

**Pull request closed without merging**